src/esp/geo/VoxelGrid.cpp

```diff
@@ -0,0 +1,923 @@
+                                    bool val,
+                                    const std::string& gridName) {
+  assert(boolGrids_.find(gridName) != boolGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  boolGrids_[gridName].get()[hashedVoxelIndex] = val;
+}
+// Getter and setter for int value voxel grids
+int VoxelGrid::getIntVoxelByIndex(const Mn::Vector3i& coords,
+                                  const std::string& gridName) {
+  assert(intGrids_.find(gridName) != intGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  return intGrids_[gridName].get()[hashedVoxelIndex];
+}
+
+void VoxelGrid::setIntVoxelByIndex(const Mn::Vector3i& coords,
+                                   int val,
+                                   const std::string& gridName) {
+  assert(intGrids_.find(gridName) != intGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  intGrids_[gridName].get()[hashedVoxelIndex] = val;
+}
+
+// Getter and setter for Float value voxel grids
+float VoxelGrid::getFloatVoxelByIndex(const Mn::Vector3i& coords,
+                                      const std::string& gridName) {
+  assert(floatGrids_.find(gridName) != floatGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  return floatGrids_[gridName].get()[hashedVoxelIndex];
+}
+
+void VoxelGrid::setFloatVoxelByIndex(const Mn::Vector3i& coords,
+                                     float val,
+                                     const std::string& gridName) {
+  assert(floatGrids_.find(gridName) != floatGrids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  floatGrids_[gridName].get()[hashedVoxelIndex] = val;
+}
+
+// Getter and setter for Vector3 value voxel grids
+Mn::Vector3 VoxelGrid::getVector3VoxelByIndex(const Mn::Vector3i& coords,
+                                              const std::string& gridName) {
+  assert(vector3Grids_.find(gridName) != vector3Grids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  return vector3Grids_[gridName].get()[hashedVoxelIndex];
+}
+
+void VoxelGrid::setVector3VoxelByIndex(const Mn::Vector3i& coords,
+                                       const Mn::Vector3& val,
+                                       const std::string& gridName) {
+  assert(vector3Grids_.find(gridName) != vector3Grids_.end());
+  int hashedVoxelIndex = hashVoxelIndex(coords);
+  vector3Grids_[gridName].get()[hashedVoxelIndex] = val;
+}
+
+std::shared_ptr<Mn::Trade::MeshData> VoxelGrid::getMeshData(
+    const std::string& gridName) {
+  if (meshDataDict_[gridName] == nullptr)
+    generateMesh(gridName);
+  return meshDataDict_[gridName];
+}
+
+Mn::GL::Mesh& VoxelGrid::getMeshGL(const std::string& gridName) {
+  if (meshDataDict_[gridName] == nullptr)
+    generateMesh(gridName);
+  return meshGLDict_[gridName];
+}
+
+Mn::Vector3 VoxelGrid::getGlobalCoords(const Mn::Vector3i& coords) {
+  Mn::Vector3 global_coords((coords[0]) * m_voxelSize[0],
```

Comment:
I'm always wary of off-by-one and similar bugs for this kind of conversion.

Would you say this is getting the world-space position of a particular corner of the voxel? Or the position of the center of the voxel?

Reply:
This gets the position of the center of the voxel.